doc/adr/0021-undo-redo.md

```diff
@@ -0,0 +1,101 @@
+
+Figma's undo implementation also lists one very important guiding principle: performing an undo followed by a redo should leave the application in the same state. This is crucial when undoing in multiplayer with multiple editors. The important implication of this principle is that the redo operation should not simply be the original undone operation, because more might have changed meanwhile, but rather the reverse of the undo operation itself. 
+
+Athens's model of time gives us a starting point for a model of undo: time only moves forward, and thus undoing an operation yield an operation that reverses the effects of the original one, to be applied in a point further in time.
+
+[Operations in Athens](doc/adr/0010-atomic-composite-grapth-operations.md) follow the principle of determinism, where the operation itself carries enough information for all participants to perform the same changes with no ambiguity over the same state. This property enables us to derive an undo operation from a state together with any operation.
+
+We can reduce the participants to three types: the undo issuer, the server, and other users. Although in theory any of these could construct the concrete undo operation, it is onerous for the server and other users to keep all possible past states that might be required for undos from anyone, for any point in time.
+
+The prime candidate to keep relevant states is the undo issuer themselves. This participant can keep a list of states needed for their own set of undoable operations. However, this operation is still optimistic in nature, and subject to conflicts. It is tantamount to saying "this is how to undo the operation given what I know now".
+
+We can also transfer the burden of resolution to the server, achieving a "true" undo since the server itself is the source of truth for the RTC system. The server would still need a way to keep or recall past states in order to resolve the undo operation. This approach would require the client issuing the undo to somehow direct the server to undo the operation, and wait for the result.
+
+
+## Approach
+
+The primary motivation for undo/redo right now is the prevention of data loss in the moment, and keeping that goal in mind we decided to pursue the first candidate (issuer resolves undo operation). Higher fidelity reconstruction of paste application state, either for inspection or data recovery, is left to future time travel functionality.
+
+For each atomic operation over a db state, we can build a corresponding atomic or composite that undoes it. For composite operations, we can compute the undo atomic/composite for each atomic operation in it, maintaining order so that the operations make sense. A redo operation is the undo operation for the effected undo.
```

Comment:
It would add to clarity if you'd mention here that undo is resolved in context of DB snapshot.
Especially when mentioning undo of undo, where resolution context is not trivial and has far reaching consequences.